keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt

```diff
@@ -0,0 +1,138 @@
+package com.netflix.spinnaker.keel.sql
+
+import com.fasterxml.jackson.module.kotlin.readValue
+import com.netflix.spinnaker.keel.api.ApiVersion
+import com.netflix.spinnaker.keel.api.ArtifactType
+import com.netflix.spinnaker.keel.api.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.name
+import com.netflix.spinnaker.keel.api.randomUID
+import com.netflix.spinnaker.keel.api.uid
+import com.netflix.spinnaker.keel.persistence.DeliveryConfigRepository
+import com.netflix.spinnaker.keel.persistence.NoSuchDeliveryConfigName
+import com.netflix.spinnaker.keel.serialization.configuredObjectMapper
+import org.jooq.DSLContext
+
+class SqlDeliveryConfigRepository(
+  private val jooq: DSLContext,
+  private val resourceTypeIdentifier: (String) -> Class<*>
+) : DeliveryConfigRepository {
+
+  private val mapper = configuredObjectMapper()
+
+  override fun store(deliveryConfig: DeliveryConfig) {
+    jooq.inTransaction {
+      val uid = randomUID().toString()
+      with(deliveryConfig) {
+        insertInto(Schema.DeliveryConfig.table)
+          .set(Schema.DeliveryConfig.uid, uid)
+          .set(Schema.DeliveryConfig.name, name)
+          .set(Schema.DeliveryConfig.application, application)
+          .onDuplicateKeyIgnore()
+          .execute()
+        artifacts.forEach { artifact ->
+          insertInto(Schema.Artifact.table)
+            .set(Schema.Artifact.uid, randomUID().toString())
+            .set(Schema.Artifact.name, artifact.name)
+            .set(Schema.Artifact.type, artifact.type.name)
+            .onDuplicateKeyIgnore()
+            .execute()
+          insertInto(Schema.DeliveryConfigArtifact.table)
+            .set(Schema.DeliveryConfigArtifact.deliveryConfigUid, uid)
+            .set(Schema.DeliveryConfigArtifact.artifactUid, select(Schema.Artifact.uid).from(Schema.Artifact.table).where(Schema.Artifact.name.eq(artifact.name)))
+            .onDuplicateKeyIgnore()
```

Comment:
I plan to put some tests around assumptions about when we can ignore / assume update for duplicate keys. It's tricky in some cases.